File: src/lib/compatibility/engine.ts
import { getComponentModel } from "@/lib/catalog/components";
import { isSafetyCriticalSlot } from "@/lib/catalog/slots";
import {
  COMPATIBILITY_RULES,
  type CompatibilityRuleDef,
} from "@/lib/compatibility/rules";
import type {
  Bike,
  BikeComponent,
  CompatibilityResult,
  CompatibilityVerdict,
  ComponentModel,
  ComponentSlot,
  EvidenceItem,
  TypedAttribute,
} from "@/types/garage";

function attrMap(attrs: TypedAttribute[]): Map<string, TypedAttribute> {
  return new Map(attrs.map((a) => [a.key, a]));
}

function resolveAttrs(
  component: BikeComponent | undefined,
  model: ComponentModel | undefined
): Map<string, TypedAttribute> {
  const merged = new Map<string, TypedAttribute>();
  if (model) {
    for (const a of model.attributes) merged.set(a.key, a);
  }
  if (component) {
    for (const a of component.attributes) merged.set(a.key, a);
  }
  return merged;
}

function getInstalled(bike: Bike, slot: ComponentSlot): BikeComponent | undefined {
  return bike.components.find((c) => c.slot === slot && !c.removedAt);
}

function readValue(
  map: Map<string, TypedAttribute>,
  key: string
): string | number | undefined {
  const a = map.get(key);
  if (!a) return undefined;
  if (a.notApplicable || a.valueEnum === "n/a" || a.valueText === "n/a") {
    return "n/a";
  }
  if (a.valueEnum !== undefined) return a.valueEnum;
  if (a.valueNum !== undefined) return a.valueNum;
  if (a.valueText !== undefined) return a.valueText;
  return undefined;
}

function missingList(
  rule: CompatibilityRuleDef,
  mapA: Map<string, TypedAttribute>,
  mapB: Map<string, TypedAttribute>
): { key: string; howToObtain: string }[] {
  const missing: { key: string; howToObtain: string }[] = [];
  for (const key of rule.requiresA) {
    if (readValue(mapA, key) === undefined) {
      missing.push({
        key: `${rule.slotA}.${key}`,
        howToObtain: rule.howToObtain[key] ?? "Herstellerdatenblatt prüfen",
      });
    }
  }
  for (const key of rule.requiresB) {
    if (readValue(mapB, key) === undefined) {
      missing.push({
        key: `${rule.slotB}.${key}`,
        howToObtain: rule.howToObtain[key] ?? "Herstellerdatenblatt prüfen",
      });
    }
  }
  return missing;
}

/** Praxis-Tabelle: Reifenbreite sollte ca. 1,4–2,4 × innere Felgenmaulweite sein */
function tireRimCompatible(tireWidthMm: number, rimInnerMm: number): boolean {
  const min = rimInnerMm * 1.4;
  const max = rimInnerMm * 2.4;
  return tireWidthMm >= min && tireWidthMm <= max;
}

function fillTemplate(
  template: string,
  a: Map<string, TypedAttribute>,
  b: Map<string, TypedAttribute>
): string {
  return template
    .replace(/\{a\.(\w+)\}/g, (_, k) => String(readValue(a, k) ?? "?"))
    .replace(/\{b\.(\w+)\}/g, (_, k) => String(readValue(b, k) ?? "?"));
}

function evaluatePredicate(
  rule: CompatibilityRuleDef,
  mapA: Map<string, TypedAttribute>,
  mapB: Map<string, TypedAttribute>
): { pass: boolean; conditional?: boolean; evidence: EvidenceItem[] } {
  const evidence: EvidenceItem[] = [];

  if (rule.predicate === "equals") {
    const keyA = rule.requiresA[0];
    const keyB = rule.requiresB[0];
    const va = readValue(mapA, keyA);
    const vb = readValue(mapB, keyB);
    const attrA = mapA.get(keyA);
    const attrB = mapB.get(keyB);
    evidence.push({
      ruleCode: rule.code,
      attributeKey: keyA,
      valueA: va,
      valueB: vb,
      sourceA: attrA?.source,
      sourceB: attrB?.source,
      verifiedAtA: attrA?.verifiedAt,
      verifiedAtB: attrB?.verifiedAt,
    });
    // n/a auf Rahmen bei Motor = kein E-Bike → Motor-Slot sollte nicht belegt sein;
    // wenn Motor fehlt, wird Regel übersprungen. Wenn beide n/a: pass.
    if (va === "n/a" && vb === "n/a") return { pass: true, evidence };
    if (va === "n/a" || vb === "n/a") return { pass: false, evidence };
    return { pass: String(va) === String(vb), evidence };
  }

  if (rule.predicate === "shock_fit") {
    const pairs: [string, string][] = [
      ["eye_to_eye_mm", "shock_eye_to_eye_mm"],
      ["stroke_mm", "shock_stroke_mm"],
      ["mount_type", "shock_mount_type"],
    ];
    let pass = true;
    for (const [ka, kb] of pairs) {
      const va = readValue(mapA, ka);
      const vb = readValue(mapB, kb);
      evidence.push({
        ruleCode: rule.code,
        attributeKey: ka,
        valueA: va,
        valueB: vb,
        sourceA: mapA.get(ka)?.source,
        sourceB: mapB.get(kb)?.source,
      });
      if (String(va) !== String(vb)) pass = false;
    }
    return { pass, evidence };
  }

  if (rule.predicate === "tire_rim_fit") {
    const tire = Number(readValue(mapA, "tire_width_mm"));
    const rim = Number(readValue(mapB, "internal_rim_width_mm"));
    evidence.push({
      ruleCode: rule.code,
      attributeKey: "tire_width_mm",
      valueA: tire,
      valueB: rim,
      note: "Fallback-Tabelle 1,4–2,4× Maulweite (ohne OEM-Freigabe → CONDITIONAL bei Pass)",
      sourceA: mapA.get("tire_width_mm")?.source,
      sourceB: mapB.get("internal_rim_width_mm")?.source,
    });
    const pass = tireRimCompatible(tire, rim);
    return { pass, conditional: pass, evidence };
  }

  if (rule.predicate === "rotor_within_max") {
    // reuse: tire width vs max clearance OR rotor vs max
    const valA = Number(
      readValue(mapA, rule.requiresA[0]) ?? readValue(mapA, "rotor_diameter_mm")
    );
    const maxB = Number(readValue(mapB, rule.requiresB[0]));
    evidence.push({
      ruleCode: rule.code,
      attributeKey: rule.requiresA[0],
      valueA: valA,
      valueB: maxB,
      sourceA: mapA.get(rule.requiresA[0])?.source,
      sourceB: mapB.get(rule.requiresB[0])?.source,
    });
    return { pass: valA <= maxB, evidence };
  }

  if (rule.predicate === "seatpost_fit") {
    const diaA = Number(readValue(mapA, "seatpost_diameter_mm"));
    const diaB = Number(readValue(mapB, "seatpost_diameter_mm"));
    const minIns = Number(readValue(mapA, "min_insertion_mm") ?? 0);
    const maxIns = Number(readValue(mapB, "max_seatpost_insertion_mm") ?? 0);
    evidence.push({
      ruleCode: rule.code,
      attributeKey: "seatpost_diameter_mm",
      valueA: diaA,
      valueB: diaB,
    });
    evidence.push({
      ruleCode: rule.code,
      attributeKey: "min_insertion_mm",
      valueA: minIns,
      valueB: maxIns,
      note: "min. Einstecktiefe ≤ max. Rahmen-Einstecktiefe",
    });
    return { pass: diaA === diaB && minIns <= maxIns, evidence };
  }

  return { pass: false, evidence };
}

function workshopHint(slots: ComponentSlot[]): string | undefined {
  if (slots.some(isSafetyCriticalSlot)) {
    return "Sicherheitsrelevante Montage: von einer Fachwerkstatt ausführen lassen. Drehmomente ausschließlich aus Herstellerdokumenten – nie schätzen.";
  }
  return undefined;
}

export function evaluateRule(
  bike: Bike,
  rule: CompatibilityRuleDef,
  candidate?: { slot: ComponentSlot; modelId: string }
): CompatibilityResult | null {
  const compA =
    candidate && candidate.slot === rule.slotA
      ? undefined
      : getInstalled(bike, rule.slotA);
  const compB =
    candidate && candidate.slot === rule.slotB
      ? undefined
      : getInstalled(bike, rule.slotB);

  const modelA =
    candidate && candidate.slot === rule.slotA
      ? getComponentModel(candidate.modelId)
      : compA?.componentModelId
        ? getComponentModel(compA.componentModelId)
        : undefined;
  const modelB =
    candidate && candidate.slot === rule.slotB
      ? getComponentModel(candidate.modelId)
      : compB?.componentModelId
        ? getComponentModel(compB.componentModelId)
        : undefined;

  // Slot A oder B nicht vorhanden und kein Kandidat → Regel nicht anwenden
  const hasA =
    !!modelA ||
    !!compA?.componentModelId ||
    (compA && (compA.attributes.length > 0 || compA.freeText));
  const hasB =
    !!modelB ||
    !!compB?.componentModelId ||
    (compB && (compB.attributes.length > 0 || compB.freeText));

  if (!hasA || !hasB) return null;

  // Free-Text ohne Katalog → INSUFFICIENT_DATA (sperrt Kompat laut Spec)
  if (
    (compA && !compA.componentModelId && !candidate) ||
    (compB && !compB.componentModelId && !(candidate && candidate.slot === rule.slotB))
  ) {
    if (
      (compA && !compA.componentModelId && rule.slotA === compA.slot) ||
      (compB && !compB.componentModelId && rule.slotB === compB.slot)
    ) {
      return {
        verdict: "INSUFFICIENT_DATA",
        ruleCode: rule.code,
        title: rule.title,
        severity: rule.severity,
        explainDe:
          "Freitext-Komponente ohne Katalogbezug – Kompatibilitätsprüfung und Setup-Automatik gesperrt (F-GAR-002).",
        missingAttributes: [
          {
            key: "component_model_id",
            howToObtain: "Katalogmodell wählen statt Freitext",
          },
        ],
        evidence: [],
        safetyWorkshopHint: workshopHint([rule.slotA, rule.slotB]),
        torqueSpecs: [...(modelA?.torqueSpecs ?? []), ...(modelB?.torqueSpecs ?? [])],
        sourceUrl: rule.sourceUrl,
      };
    }
  }

  const mapA = resolveAttrs(
    candidate?.slot === rule.slotA ? undefined : compA,
    modelA
  );
  const mapB = resolveAttrs(
    candidate?.slot === rule.slotB ? undefined : compB,
    modelB
  );

  const missing = missingList(rule, mapA, mapB);
  if (missing.length > 0) {
    return {
      verdict: "INSUFFICIENT_DATA",
      ruleCode: rule.code,
      title: rule.title,
      severity: rule.severity,
      explainDe: `Fehlende Attribute – Engine rät nicht (F-GAR-003.1): ${missing
        .map((m) => m.key)
        .join(", ")}`,
      missingAttributes: missing,
      evidence: [],
      safetyWorkshopHint: workshopHint([rule.slotA, rule.slotB]),
      torqueSpecs: [...(modelA?.torqueSpecs ?? []), ...(modelB?.torqueSpecs ?? [])],
      sourceUrl: rule.sourceUrl,
    };
  }

  const { pass, conditional, evidence } = evaluatePredicate(rule, mapA, mapB);
  let verdict: CompatibilityVerdict;
  if (pass) {
    verdict =
      conditional || rule.onPass === "CONDITIONAL" ? "CONDITIONAL" : "COMPATIBLE";
  } else {
    verdict = rule.onFail;
  }

  return {
    verdict,
    ruleCode: rule.code,
    title: rule.title,
    severity: rule.severity,
    conditionText: verdict === "CONDITIONAL" ? rule.conditionText : undefined,
    explainDe: pass
      ? `${rule.title}: Prüfung bestanden.`
      : fillTemplate(rule.explainFailDe, mapA, mapB),
    missingAttributes: [],
    evidence,
    safetyWorkshopHint: workshopHint([rule.slotA, rule.slotB]),
    torqueSpecs: [...(modelA?.torqueSpecs ?? []), ...(modelB?.torqueSpecs ?? [])],
    sourceUrl: rule.sourceUrl,
  };
}

/** Aggregiertes Bike-Urteil: schlechtestes Einzelergebnis gewinnt */
export function aggregateVerdict(
  results: CompatibilityResult[]
): CompatibilityVerdict {
  if (results.some((r) => r.verdict === "INCOMPATIBLE")) return "INCOMPATIBLE";
  if (results.some((r) => r.verdict === "INSUFFICIENT_DATA"))
    return "INSUFFICIENT_DATA";
  if (results.some((r) => r.verdict === "CONDITIONAL")) return "CONDITIONAL";
  if (results.length === 0) return "INSUFFICIENT_DATA";
  return "COMPATIBLE";
}

export function checkBikeCompatibility(bike: Bike): CompatibilityResult[] {
  const results: CompatibilityResult[] = [];
  for (const rule of COMPATIBILITY_RULES) {
    const r = evaluateRule(bike, rule);
    if (r) results.push(r);
  }
  return results;
}

export function checkCandidateOnBike(
  bike: Bike,
  slot: ComponentSlot,
  modelId: string
): CompatibilityResult[] {
  const results: CompatibilityResult[] = [];
  for (const rule of COMPATIBILITY_RULES) {
    if (rule.slotA !== slot && rule.slotB !== slot) continue;
    const r = evaluateRule(bike, rule, { slot, modelId });
    if (r) results.push(r);
  }
  return results;
}

export function verdictLabel(v: CompatibilityVerdict): string {
  switch (v) {
    case "COMPATIBLE":
      return "Kompatibel";
    case "CONDITIONAL":
      return "Bedingt";
    case "INCOMPATIBLE":
      return "Inkompatibel";
    case "INSUFFICIENT_DATA":
      return "Daten fehlen";
  }
}

export function verdictColorClass(v: CompatibilityVerdict): string {
  switch (v) {
    case "COMPATIBLE":
      return "bg-success/20 text-success border-success/40";
    case "CONDITIONAL":
      return "bg-warning/20 text-warning border-warning/40";
    case "INCOMPATIBLE":
      return "bg-error/20 text-error border-error/40";
    case "INSUFFICIENT_DATA":
      return "bg-muted text-text-secondary border-border";
  }
}

export { attrMap };

File: mobile/lib/domain/compatibility/engine.dart
/// Kompat-Engine — Spiegel src/lib/compatibility/engine.ts (ohne Katalog-Modelle).
/// Attribute kommen aus BikeComponent.attributes; fehlende Keys → INSUFFICIENT_DATA.

import '../component.dart';
import 'rules.dart';

dynamic _read(Map<String, dynamic> attrs, String key) {
  if (!attrs.containsKey(key)) return null;
  final v = attrs[key];
  if (v == null) return null;
  if (v == 'n/a') return 'n/a';
  return v;
}

bool _tireRimCompatible(num tireWidthMm, num rimInnerMm) {
  final min = rimInnerMm * 1.4;
  final max = rimInnerMm * 2.4;
  return tireWidthMm >= min && tireWidthMm <= max;
}

String _fill(
  String template,
  Map<String, dynamic> a,
  Map<String, dynamic> b,
) {
  return template
      .replaceAllMapped(
        RegExp(r'\{a\.(\w+)\}'),
        (m) => '${_read(a, m[1]!) ?? '?'}',
      )
      .replaceAllMapped(
        RegExp(r'\{b\.(\w+)\}'),
        (m) => '${_read(b, m[1]!) ?? '?'}',
      );
}

({bool pass, bool conditional}) _evaluatePredicate(
  CompatibilityRuleDef rule,
  Map<String, dynamic> mapA,
  Map<String, dynamic> mapB,
) {
  switch (rule.predicate) {
    case 'equals':
      final va = _read(mapA, rule.requiresA.first);
      final vb = _read(mapB, rule.requiresB.first);
      if (va == 'n/a' && vb == 'n/a') return (pass: true, conditional: false);
      if (va == 'n/a' || vb == 'n/a') return (pass: false, conditional: false);
      return (pass: '$va' == '$vb', conditional: false);
    case 'shock_fit':
      const pairs = [
        ('eye_to_eye_mm', 'shock_eye_to_eye_mm'),
        ('stroke_mm', 'shock_stroke_mm'),
        ('mount_type', 'shock_mount_type'),
      ];
      var pass = true;
      for (final (ka, kb) in pairs) {
        if ('${_read(mapA, ka)}' != '${_read(mapB, kb)}') pass = false;
      }
      return (pass: pass, conditional: false);
    case 'tire_rim_fit':
      final tire = num.tryParse('${_read(mapA, 'tire_width_mm')}') ?? 0;
      final rim = num.tryParse('${_read(mapB, 'internal_rim_width_mm')}') ?? 0;
      final pass = _tireRimCompatible(tire, rim);
      return (pass: pass, conditional: pass);
    case 'rotor_within_max':
      final valA =
          num.tryParse('${_read(mapA, rule.requiresA.first)}') ?? 0;
      final maxB =
          num.tryParse('${_read(mapB, rule.requiresB.first)}') ?? 0;
      return (pass: valA <= maxB, conditional: false);
    case 'seatpost_fit':
      final diaA =
          num.tryParse('${_read(mapA, 'seatpost_diameter_mm')}') ?? -1;
      final diaB =
          num.tryParse('${_read(mapB, 'seatpost_diameter_mm')}') ?? -2;
      final minIns =
          num.tryParse('${_read(mapA, 'min_insertion_mm')}') ?? 0;
      final maxIns =
          num.tryParse('${_read(mapB, 'max_seatpost_insertion_mm')}') ?? 0;
      return (pass: diaA == diaB && minIns <= maxIns, conditional: false);
    default:
      return (pass: false, conditional: false);
  }
}

BikeComponent? _installed(List<BikeComponent> comps, ComponentSlot slot) {
  for (final c in comps) {
    if (c.slot == slot && c.isInstalled) return c;
  }
  return null;
}

String? _workshopHint(CompatibilityRuleDef rule) {
  if (rule.severity == RuleSeverity.safetyCritical) {
    return 'Sicherheitsrelevante Montage: Fachwerkstatt. Drehmomente nur aus Herstellerdokumenten.';
  }
  return null;
}

CompatibilityResult? evaluateRule(
  List<BikeComponent> components,
  CompatibilityRuleDef rule,
) {
  final compA = _installed(components, rule.slotA);
  final compB = _installed(components, rule.slotB);
  if (compA == null || compB == null) return null;

  // Freitext ohne Attribute → INSUFFICIENT_DATA (Spec: nicht raten)
  if (compA.attributes.isEmpty && compA.catalogModelId == null ||
      compB.attributes.isEmpty && compB.catalogModelId == null) {
    // still allow if attributes present on both
  }

  final mapA = Map<String, dynamic>.from(compA.attributes);
  final mapB = Map<String, dynamic>.from(compB.attributes);

  final missing = <({String key, String howToObtain})>[];
  for (final key in rule.requiresA) {
    if (_read(mapA, key) == null) {
      missing.add((
        key: '${rule.slotA.apiId}.$key',
        howToObtain: rule.howToObtain[key] ?? 'Herstellerdatenblatt prüfen',
      ));
    }
  }
  for (final key in rule.requiresB) {
    if (_read(mapB, key) == null) {
      missing.add((
        key: '${rule.slotB.apiId}.$key',
        howToObtain: rule.howToObtain[key] ?? 'Herstellerdatenblatt prüfen',
      ));
    }
  }
  if (missing.isNotEmpty) {
    return CompatibilityResult(
      verdict: CompatVerdict.insufficientData,
      ruleCode: rule.code,
      title: rule.title,
      severity: rule.severity,
      explainDe:
          'Fehlende Attribute — kein COMPATIBLE ohne vollständige Faktenlage.',
      missingAttributes: missing,
      safetyWorkshopHint: _workshopHint(rule),
      sourceUrl: rule.sourceUrl,
    );
  }

  final eval = _evaluatePredicate(rule, mapA, mapB);
  if (eval.pass) {
    return CompatibilityResult(
      verdict: eval.conditional
          ? CompatVerdict.conditional
          : rule.onPass,
      ruleCode: rule.code,
      title: rule.title,
      severity: rule.severity,
      explainDe: eval.conditional
          ? (rule.conditionText ?? 'Bedingt kompatibel')
          : 'Regel erfüllt.',
      conditionText: rule.conditionText,
      safetyWorkshopHint: _workshopHint(rule),
      sourceUrl: rule.sourceUrl,
    );
  }
  return CompatibilityResult(
    verdict: rule.onFail,
    ruleCode: rule.code,
    title: rule.title,
    severity: rule.severity,
    explainDe: _fill(rule.explainFailDe, mapA, mapB),
    conditionText: rule.conditionText,
    safetyWorkshopHint: _workshopHint(rule),
    sourceUrl: rule.sourceUrl,
  );
}

List<CompatibilityResult> checkBikeCompatibility(
  List<BikeComponent> components,
) {
  final out = <CompatibilityResult>[];
  for (final rule in compatibilityRules) {
    final r = evaluateRule(components, rule);
    if (r != null) out.add(r);
  }
  return out;
}

/// Kandidat (z. B. Shop-Produkt) gegen installierte Teile prüfen.
/// Ersetzt den Slot des Kandidaten und wertet nur betroffene Regeln aus.
List<CompatibilityResult> checkCandidateOnBike(
  List<BikeComponent> installed,
  BikeComponent candidate,
) {
  final comps = <BikeComponent>[
    for (final c in installed)
      if (c.slot != candidate.slot) c,
    candidate,
  ];
  final out = <CompatibilityResult>[];
  for (final rule in compatibilityRules) {
    if (rule.slotA != candidate.slot && rule.slotB != candidate.slot) {
      continue;
    }
    final r = evaluateRule(comps, rule);
    if (r != null) out.add(r);
  }
  return out;
}

CompatVerdict aggregateVerdict(List<CompatibilityResult> results) {
  if (results.isEmpty) return CompatVerdict.insufficientData;
  var worst = CompatVerdict.compatible;
  for (final r in results) {
    worst = _worse(worst, r.verdict);
  }
  return worst;
}

CompatVerdict _worse(CompatVerdict a, CompatVerdict b) {
  int rank(CompatVerdict v) => switch (v) {
        CompatVerdict.incompatible => 3,
        CompatVerdict.insufficientData => 2,
        CompatVerdict.conditional => 1,
        CompatVerdict.compatible => 0,
      };
  return rank(b) > rank(a) ? b : a;
}

String verdictLabel(CompatVerdict v) => switch (v) {
      CompatVerdict.compatible => 'Kompatibel',
      CompatVerdict.conditional => 'Bedingt',
      CompatVerdict.incompatible => 'Inkompatibel',
      CompatVerdict.insufficientData => 'Daten fehlen',
    };
